src/modules/convex/task/mlp.hpp

```diff
@@ -0,0 +1,334 @@
+
+    static void gradientInPlace(
+            model_type                          &model,
+            const independent_variables_type    &y,
+            const dependent_variable_type       &z,
+            const double                        &stepsize);
+
+    static double loss(
+            const model_type                    &model,
+            const independent_variables_type    &y,
+            const dependent_variable_type       &z);
+
+    static ColumnVector predict(
+            const model_type                    &model,
+            const independent_variables_type    &y,
+            const bool                          get_class);
+
+    const static int RELU = 0;
+    const static int SIGMOID = 1;
+    const static int TANH = 2;
+
+    static double sigmoid(const double &xi) {
+        return 1. / (1. + std::exp(-xi));
+    }
+
+    static double relu(const double &xi) {
+        return xi*(xi>0);
+    }
+
+    static double tanh(const double &xi) {
+        return std::tanh(xi);
+    }
+
+
+private:
+
+    static double sigmoidDerivative(const double &xi) {
+        double value = sigmoid(xi);
+        return value * (1. - value);
+    }
+
+    static double reluDerivative(const double &xi) {
+        return xi>0;
+    }
+
+    static double tanhDerivative(const double &xi) {
+        double value = tanh(xi);
+        return 1-value*value;
+    }
+
+    static void feedForward(
+            const model_type                    &model,
+            const independent_variables_type    &y,
+            std::vector<ColumnVector>           &net,
+            std::vector<ColumnVector>           &x);
+
+    static void endLayerDeltaError(
+            const std::vector<ColumnVector>     &net,
+            const std::vector<ColumnVector>     &x,
+            const dependent_variable_type       &z,
+            ColumnVector                        &delta_N);
+
+    static void errorBackPropagation(
+            const ColumnVector                  &delta_N,
+            const std::vector<ColumnVector>     &net,
+            const model_type                    &model,
+            std::vector<ColumnVector>           &delta);
+};
+
+template <class Model, class Tuple>
+void
+MLP<Model, Tuple>::gradientInPlace(
+        model_type                          &model,
+        const independent_variables_type    &y,
+        const dependent_variable_type       &z,
+        const double                        &stepsize) {
+    (void) model;
+    (void) z;
+    (void) y;
+    (void) stepsize;
+    std::vector<ColumnVector> net;
+    std::vector<ColumnVector> x;
+    std::vector<ColumnVector> delta;
+    ColumnVector delta_N;
+
+    feedForward(model, y, net, x);
+    endLayerDeltaError(net, x, z, delta_N);
+    errorBackPropagation(delta_N, net, model, delta);
+
+    uint16_t N = model.u.size(); // assuming nu. of layers >= 1
+    uint16_t k, s, j;
+    // #TODO fix lambda
```

Comment:
We will use a different approach, making use of the convex architecture.  So I will remove this altogether.